### PR TITLE
News Downloader - feed config editor

### PR DESCRIPTION
**Pull request by mwoz123**
zrzut ekranu z 2018-07-30 23-12-21

No validation yet - in next phase.

### CODE REVIEW

**Comment by Frenzie** [comment]
At least pcall loadstring or something. ;-)

File: plugins/newsdownloader.koplugin/main.lua [hunk 1]
```diff
@@ -388,6 +385,54 @@ function NewsDownloader:setCustomDownloadDirectory()
     }:chooseDir()
 end
 
+function NewsDownloader:changeFeedConfig() 
+    local feed_config_file = io.open(feed_config_path, "rb")
+    local config = feed_config_file:read("*all")
+    feed_config_file:close()
+    config_editor = InputDialog:new{ --todo change to local var, requires fix in save and cancel. No idea how.
```
Comment (poire-z) [inline plugins/newsdownloader.koplugin/main.lua]: You need to define the local variable first, so it exists and you can reference it then:
```lua
local config_editor
config_editor = InputDialog:new{ [...] -- you can then reference config_editor here in the callbacks }
```

File: plugins/newsdownloader.koplugin/main.lua [hunk 2]
```diff
@@ -388,6 +385,54 @@ function NewsDownloader:setCustomDownloadDirectory()
...
+        title = _("Feed config editor"),
+        input = config,
+        input_type = "string",
+        fullscreen = true,
+        condensed = true,
+        allow_newline = true,
+        cursor_at_end = false,
+        add_scroll_buttons = true,
+        add_nav_bar = true,
```
Comment (poire-z) [inline plugins/newsdownloader.koplugin/main.lua]: You don't need both, choose add_scroll_buttons OR add_nav_bar .

File: plugins/newsdownloader.koplugin/main.lua (hunk 1 above)
Comment (mwoz123) [reply plugins/newsdownloader.koplugin/main.lua]: never would thought spiting in-line declaration and assignment can make any difference :O
Thanks:)

File: plugins/newsdownloader.koplugin/main.lua (hunk 2 above)
Comment (mwoz123) [reply plugins/newsdownloader.koplugin/main.lua]: ok. thanks

**Comment by mwoz123** [comment]
>At least pcall loadstring or something. ;-)

you mean something similar to that:
https://github.com/koreader/koreader/blob/e37c34531ad3327f5602ae8efdaab23928ae1cac/plugins/evernote.koplugin/slt2.lua#L32
?

**Comment by NiLuJe** [comment]
@mwoz123 : Nope, pcall (grep a bit, you'll find a bunch of them around ;)).

assert will probably behave like every assert should: abort if false, wihich is probably not what you want ;).

**Comment by Frenzie** [comment]
@mwoz123 

> you mean something similar to that:

I'm not really sure what that's supposed to be doing tbh. That sounds more like you should use `pcall(dofile(filename))` but I'm not sure what that'd be good for.

```lua
-- saving routine
local syntax_totes_okay = pcall(loadstring("user_supplied_string"))
if syntax_totes_okay then
  -- save
else
  -- warn
end
```

File: plugins/newsdownloader.koplugin/main.lua [hunk 3]
```diff
@@ -379,15 +376,63 @@ function NewsDownloader:setCustomDownloadDirectory()
...
+                            end
+                        })
+                    end,
+                },
+            }
+        },
+    }
+    UIManager:show(config_editor)
+    config_editor:onShowKeyboard()
+
```
Comment (Frenzie) [inline plugins/newsdownloader.koplugin/main.lua]: stray

File: plugins/newsdownloader.koplugin/main.lua [hunk 4]
```diff
@@ -379,15 +376,70 @@ function NewsDownloader:setCustomDownloadDirectory()
...
+                    text = _("Cancel"),
+                    callback = function()
+                        UIManager:close(config_editor)
+                    end,
+                },
+                {
+                    text = _("Save"),
+                    callback = function()
+                        local new_config = config_editor:getInputText()
+                        local syntax_totes_okay, error = pcall(loadstring(new_config))
```
Comment (Frenzie) [inline plugins/newsdownloader.koplugin/main.lua]: Please don't use that as an actual variable name. :rofl:

**Comment by mwoz123** [comment]
Any thing else?

**Comment by poire-z** [comment]
I'll be adding this weekend some more options to InputDialog that you can benefit from.
Instead of your buttons, you could just provide a `save_callback = function(content)` and do your saving here. A Save and Close buttons will be added with all necessary check so you don't have to bother with them.

So, if you can wait, with very small changes to this PR, you'll be making an other simple example of how to use that (the Text Editor plugin being a more complex example).

**Comment by mwoz123** [comment]
ok. please let me know, once finished. thanks

**Comment by poire-z** [comment]
OK, once #4135 is merged, you should be able to remove your `buttons = {...}` and use instead:
```lua
-- Configuration saving callback
save_callback = function(content)
    if content and #content > 0 then
        local parse_error = util.checkLuaSyntax(content)
        if not parse_error then
            feed_config_file = io.open(feed_config_path, "w")
            feed_config_file:write(content)
            feed_config_file:close()
            return true, _("Configuration saved")
        else
            return false, T(_("Configuration invalid: %1"), parse_error)
        end
    end
    return false, _("Configuration empty")
end,
```
(untested)
And if you want/need, a `reset_callback` to add a Reset button and restore the current configuration.

**Comment by poire-z** [comment]
#4135 merged, so you can try this modification.

**Comment by mwoz123** [comment]
@poire-z Adapted to changes. Code changes you pasted works without any problems.

It only looks as if there might be a small issue in `util.checkLuaSyntax(content)` as it successfully validates when I change config first line e.g. to `zzzzreturn {`

**Comment by poire-z** [comment]
`zzzzreturn {` fails with `unexpected symbol near '<eof'>`

`zzzzreturn {}` indeed does not fail. But it's valid lua syntax, it's like `zzzreturn()`:
```
$ lua
Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio
> zzzreturn {}
stdin:1: attempt to call global 'zzzreturn' (a nil value)
stack traceback:
        stdin:1: in main chunk
        [C]: ?
>
```

`util.checkLuaSyntax` actually does only the `loadstring` part of your original `pcall(loadstring(new_config))`.
It can't do more, as **running** (with `pcall`) the code can actually `do` stuff :)

But your plugin can do what you did origianlly as a 2nd step : `pcall(loadstring(new_config))` if you're sure you need it, as you expect it to `return something`
(But if the user type `os.remove("reader.lua")`, it will be run.)
You decide.

**Comment by poire-z** [comment]
What happens if one just put `return 42` ? which will survive your pcall, but may fail your plugin if it expects it to be a table.

**Comment by mwoz123** [comment]
Validation initially was not part of this pr. It iwas for next time.but then "pcall" validation was proposed. 

Until now user had no validation at all and had to edit in external editor. 

 I would prefer to do bigger validation next time.

**Comment by Frenzie** [comment]
My point about super basic syntax validation is that it'd take 10 seconds to write so that there was no reason to leave that out.

**Comment by mwoz123** [comment]
@Frenzie and so basic pcall validation was  implemented. 
For more  it will require some addional time, that I'd like to do next time. 
> but may fail your plugin if it expects it to be a table.

For current config `type(content)` == string (not table), so it won't be easy nor quick for me to do it. 

@poire-z Can it be done next time?

**Comment by poire-z** [comment]
Not sure it's even required, I was just making a point that if pure parser syntax check wasn't enough that you feel you needed to pcall it, well, pcall may too not be enough :)

**Pull request merged**